src/server/game/Server/Packets/BattlePetPackets.cpp

```diff
@@ -0,0 +1,165 @@
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "BattlePetPackets.h"
+
+ByteBuffer& operator<<(ByteBuffer& data, WorldPackets::BattlePet::BattlePetSlot const& slot)
+{
+    data << (slot.Pet.Guid.IsEmpty() ? ObjectGuid::Create<HighGuid::BattlePet>(0) : slot.Pet.Guid);
+    data << uint32(0); // Collar ID - what does it mean?
+    data << slot.Index;
+    data.WriteBit(slot.Locked);
+    data.FlushBits();
+
+    return data;
+}
+
+ByteBuffer& operator<<(ByteBuffer& data, WorldPackets::BattlePet::BattlePet const& pet)
+{
+    data << pet.Guid;
+    data << pet.Species;
+    data << pet.CreatureID;
+    data << uint32(0); // Collar ID - what does it mean?
+    data << pet.Breed;
+    data << pet.Level;
+    data << pet.Exp;
+    data << pet.Flags;
+    data << pet.Power;
+    data << pet.Health;
+    data << pet.MaxHealth;
+    data << pet.Speed;
+    data << pet.Quality;
+    data.WriteBits(pet.Name.size(), 7);
+    data.WriteBit(!pet.Owner.IsEmpty()); // HasOwnerInfo
+    data.WriteBit(pet.Name.empty()); // NoRename
+    data.FlushBits();
+
+    if (!pet.Owner.IsEmpty())
+    {
+        data << pet.Owner;
+        data << pet.Owner.GetRealmId(); // Virtual
+        data << pet.Owner.GetRealmId(); // Native
+    }
+
+    data.WriteString(pet.Name);
+
+    return data;
+}
+
+WorldPacket const* WorldPackets::BattlePet::BattlePetJournal::Write()
+{
+    _worldPacket << Trap;
+    _worldPacket << Slots.size();
+    _worldPacket << Pets.size();
+
+    for (auto const& slot : Slots)
+        _worldPacket << slot;
+
+    for (auto const& pet : Pets)
+        _worldPacket << pet;
+
+    _worldPacket.WriteBit(1); // HasJournalLock
+    _worldPacket.FlushBits();
+
+    return &_worldPacket;
+}
+
+WorldPacket const* WorldPackets::BattlePet::BattlePetUpdates::Write()
+{
+    _worldPacket << Pets.size();
+
+    for (auto const& pet : Pets)
+        _worldPacket << pet;
+
+    _worldPacket.WriteBit(PetAdded);
+    _worldPacket.FlushBits();
+
+    return &_worldPacket;
+}
+
+
+WorldPacket const* WorldPackets::BattlePet::PetBattleSlotUpdates::Write()
+{
+    _worldPacket << Slots.size();
+
+    for (auto const& slot : Slots)
+        _worldPacket << slot;
+
+    _worldPacket.WriteBit(NewSlot);
+    _worldPacket.WriteBit(AutoSlotted);
+    _worldPacket.FlushBits();
+
+    return &_worldPacket;
+}
+
+void WorldPackets::BattlePet::BattlePetSetBattleSlot::Read()
+{
+    _worldPacket >> PetGuid;
+    _worldPacket >> Slot;
+}
+
+void WorldPackets::BattlePet::BattlePetModifyName::Read()
+{
+    _worldPacket >> PetGuid;
+    uint32 nameLength = _worldPacket.ReadBits(7);
+    bool hasDeclinedNames = _worldPacket.ReadBit();
+    Name = _worldPacket.ReadString(nameLength);
+
+    if (hasDeclinedNames)
+    {
+        uint8 declinedNameLengths[5];
```

Comment:
These are only lengths, not the names. For strings I use variable <code>DeclinedName Declined;</code>